sklearn/impute/_iterative.py

```diff
@@ -756,8 +756,17 @@ def fit_transform(self, X, y=None, **params):
             self.n_iter_ = 0
             return super()._concatenate_indicator(Xt, X_indicator)
 
-        self._min_value = self._validate_limit(self.min_value, "min", X.shape[1])
-        self._max_value = self._validate_limit(self.max_value, "max", X.shape[1])
+        self._min_value = self._validate_limit(
+            self.min_value, "min", complete_mask.shape[1]
+        )
+        self._max_value = self._validate_limit(
+            self.max_value, "max", complete_mask.shape[1]
+        )
```

Comment:
I was wondering if instead of checking the shape of the limit in [lines 681-686](https://github.com/scikit-learn/scikit-learn/blob/2b2e2903e5635dd93a741c955a87260fb69cfc3d/sklearn/impute/_iterative.py#L681) we could do the validation in the beginning of `fit() `and `fit_transform()`, like we usually do. Something like `if len(self.min_value) != X.shape[1]: raise ValueError(...)` ... It feels a bit cleaner to me.

This would mean that the code in li. 759-764 could stay as it was before.

But I think your take on dealing with this is also okay.

Reply:
Thanks for the comments! It makes sense to me if we can validate their shapes during `fit` and `fit_transform`. Ultimately, I understand that we just want to check the length of the limits whether they align with `X` and safely convert them to arrays with `check_array`.

What about moving these lines (`_validate_limit` calls and assignments) to `fit` and `fit_transform`? Then here we could basically apply the mask to `self._min_value` and `self._max_value`.

Reply:
Thanks for the comments! As @lesteve suggested above, it could be problematic when either `X`, `self.min_value`, or `self.max_value` are not `np.ndarray`. We might have to safely convert all of them to `np.ndarray`s first then we can check the lengths. However, this seems redundant because we call `_validate_data(X)` inside `self._initial_imputation` and `check_array(self.min_value)` inside `self._validate_limit`. Which approach do you think would work best?

Reply:
I think both is alright, but I agree that earlier validation is more readable.

I tried this locally:

1. revert changes in li. 759-764 (starting with `self._min_value = self._validate_limit(`)

2. insert into li. 747 (right before `super()._fit_indicator(complete_mask)`):
```python
        if len(self.min_value) != complete_mask.shape[1]: 
            raise ValueError(
                f"'min_value' should be of shape ({complete_mask.shape[1]},)"
                f"when an array-like is provided. Got {len(self.min_value)}, instead."
            )
        
        if len(self.max_value) != complete_mask.shape[1]: 
            raise ValueError(
                f"'max_value' should be of shape ({complete_mask.shape[1]},)"
                f"when an array-like is provided. Got {len(self.max_value)}, instead."
            )
```

3. outcomment / delete li. 681-686 (starting with `if not limit.shape[0] == n_features:`)

This seems to work well and looks a bit neater (if we straighten the two "raise ValueErrors" into one check).

What do you think?

Reply:
Thank you for your suggestions! I found two other issues considering this solution but managed to resolve them in my recent push.

- `self.min_value`/`self.max_value` can be `None` or a scalar. We may need to replace the condition with `self.min_value is not None and not np.isscalar(self.min_value) and len(self.min_value) != complete_mask.shape[1]` instead. Personally, I prefer letting `_validate_limit` handle these checks implicitly, but I also agree that passing `complete_mask` to `_validate_limit` is somewhat unclear.
- When `self.min_value` is a scalar, after `_validate_limit` the `self._min_value` variable already had the same length as `X`. In this case, we need to conditionally index only if `self.min_value` is an array-like since the beginning.

Please let me know if we could further enhance the code readability.